### ISSUE TITLE
Bitwig 3.3 / State loading

### ISSUE THREAD

**Issue by abique** [robbert-vdh/yabridge#43]
Hi,

I'm testing yabridge for the first time and I think it is a great software.

Here is a problematic scenario maybe:

- Bitwig loads a state (on the main thread)
- Then while it is loading, we receive audioMasterAutomate, but on a random thread and while load state is not yet finished on the main thread

It should not be necessary as after loading the preset, Bitwig will of course scan all the parameters.
Then bitwig will write a line in the log saying that the plugin should not call audioMasterAutomate while loading a preset, and while doing that it will query the plugin name and vendor name which somehow never returns from your side.

I've fixed that by caching those two strings during plugin initialization.

Regards,
Alex

**Comment by abique** [comment]
Fixed for bitwig 3.3 beta 2.

**Comment by robbert-vdh** [comment]
Hey, thanks a lot for the heads up! Yeah I also thought it was weird that plugins would call `audioMasterAutomate()` while loading preset state. My guess is that they're reusing some internal parameter setting functions that are also used for GUI interaction.

The reason why this was affecting yabridge (and likely also other VST bridges if they're implemented in a similar way), is because at the moment yabridge can only do a single `dispatch()` and a single `audioMaster()` call at a time because all communication happens over simple long lived UNIX domain sockets. Basically, this was the situation with Bitwig Studio 3.3 beta 1:

```
   Thread  Socket  Bitwig                   yabridge     Serum
   ...
1) bw1     1       effSetChunk()        --> ...      -->
2)  wine1  2                            <-- ...      <-- audioMasterUpdateDisplay()
3)   ya1   2       <returns 1>          --> ...      -->
4)  wine1  2                            <-- ...      <-- audioMasterBeginEdit()
5)   ya1   2       <returns 1>          --> ...      -->
6)  wine1  2                            <-- ...      <-- audioMasterAutomate()
7) bw2     1       effGetVendorString() --> ...      -->
```

In Bitwig 3.3 beta 1 the call on `6)` would only return after the plugin has finished loading. This in turn would cause the whole call to `effSetChunk()` to not return until after that point. Because all `dispatch()` calls are going over the same socket, the `effGetVendorString()` call on `7)` will thus wait for `6)` to return which results in a fun deadlock.

I've been wanting to come up with a solution that will spawn new sockets when necessary to work around these situations, but I haven't gotten around to doing so and I also don't want to overcomplicate things too much. There was a similar issue with nested `dispatch()` calls in REAPER that would also benefit from such a change #29 (there's currently a workaround behind an option, but that's not a great solution for the long term).

Again, thanks a lot for looking at and fixing the issue!

**Comment by abique** [comment]
I think sooner or later you will need recursive calls.

**Comment by robbert-vdh** [comment]
Yeah I agree! It's just that I'd never seen mutual recursion happen anywhere, so I hadn't taken it into consideration when designing yabridge's communication model.

**Comment by robbert-vdh** [comment]
I spent the last few days rewriting yabridge's communication model to be asynchronous and to spawn additional threads as needed in [this branch](https://github.com/robbert-vdh/yabridge/tree/feature/socket-rework), and I can confirm that these changes fix the state loading issues with Bitwig Studio 3.3 beta 1.

**Comment by abique** [comment]
This is amazing!
Well done!

**Comment by abique** [comment]
By the way do you have plans to support vst3?

**Comment by abique** [comment]
Also as a side note, when testing plugins with Bitwig you can check the logs at ~/.BitwigStudio/log/engine.log.
If we detect a behavior that is not correct to our understanding of the spec, we log it. Search for "misbehaving" in the logs.

**Comment by robbert-vdh** [comment]
I do want to support VST3 eventually, but I've been holding off on it since it's so incredibly complex compared to VST2 and because it's probably going to be a ton of additional work since it has nothing in common with the VST2 format. Right now this is fine because most plugin manufacturers still offer VST2 support, but since Steinberg has stopped giving out licenses this will probably change at some point. This communication rewrite has been in the back of my head for a while now and it was the last big change I wanted to make to yabridge before looking at VST3 support, so I'll probably end up working on that in the coming few months when I have some spare time.

**Comment by robbert-vdh** [comment]
> Also as a side note, when testing plugins with Bitwig you can check the logs at ~/.BitwigStudio/log/engine.log.
> If we detect a behavior that is not correct to our understanding of the spec, we log it. Search for "misbehaving" in the logs.

Oh, that's really good to know, thanks! I did some digging before and I ended up using `bitwig-studio --log-dir /tmp/bitwig-logs -DlogLevel=trace` when I needed to take a look at the engine logs (and before I figured that out I used `strace` :sweat_smile:), but I didn't realize those logs were always being written somewhere which is much easier of course, thanks!

**Comment by abique** [comment]
VST3 is not as complicated as it seems.

**Comment by robbert-vdh** [comment]
In case you're interested (since you've asked about this before), yabridge now supports VST3 plugin bridging as of yabridge 3.0.0! Funnily enough Bitwig was the only DAW where I did not run into any breaking VST3 implementation bugs or unexpected quirks. Although I did manage to report a relatively minor bug with in-engine plugin hosting (where VST3 modules wouldn't be unloaded even after closing the project and would thus leak resources) and a missing feature in the VST3 implementation that has been implemented for VST2 plugins (pitch labels), so there was at least something I could report haha.
